### PR TITLE
Add support to shrink the number of shards of a table or partition

### PR DESCRIPTION
**Pull request by mfussenegger**
- [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 1]
```diff
@@ -202,6 +261,220 @@ private Session session() {
         return result;
     }
 
+    private CompletableFuture<Long> executeAlterTableChangeNumberOfShards(AlterTableAnalyzedStatement analysis) {
+        final DocTableInfo table = analysis.table();
+        final boolean isPartitioned = table.isPartitioned();
+        String sourceIndexName, sourceIndexAlias;
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: haven't seen this in a while :), although it's ok syntactically, we don't generally define multiple variables in one go

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 2]
```diff
@@ -202,6 +261,220 @@ private Session session() {
...
+    }
+
+    private static void validateForResizeRequest(ClusterState clusterState, String indexName, int targetNumberOfShards) {
+        final IndexMetaData indexMetaData = clusterState.metaData().index(indexName);
+        validateNumberOfShardsForResize(indexMetaData, targetNumberOfShards);
+        validateReadOnlyIndexForResize(indexMetaData);
+    }
+
+    @VisibleForTesting
+    static Integer getValidNumberOfShards(final Settings settings) {
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: why "ValidNumberOfShards"? (to me it seems as it's just reading the setting)

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 3]
```diff
@@ -202,6 +261,220 @@ private Session session() {
...
+            });
+
+        transportResizeAction.execute(request, listener);
+        return listener;
+    }
+
+    private CompletableFuture<Long> deleteIndex(String indexName) {
+        DeleteIndexRequest request = new DeleteIndexRequest(indexName);
+
+        FutureActionListener<AcknowledgedResponse, Long> listener = new FutureActionListener<>(r -> 0L);
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: do we care about a possible failure of the delete operation? ( I'm thinking of deleting the backup as for the shrinked leftover we have an operation to clean up)

Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: if the operation is retried, the backup is also deleted before the new shrink begins.
In any case, the backup concept will go in #7992

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java (hunk 2 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: Renamed it in https://github.com/crate/crate/pull/7992/files#diff-b5f2ccdb9fb89e66959fddf9c1a0cec2R343

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java (hunk 1 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: what's wrong with that?  :)

I'll change it on master after merge - promise

Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: If we don't want to have multiple variable definitions in a single line we should see if there is a checkstyle rule for that and enable it.

**Pull request merged**